httpbp/response.go

```diff
@@ -1,313 +1,173 @@
 package httpbp
 
 import (
-	"context"
 	"encoding/json"
-	"errors"
 	"fmt"
 	"html/template"
+	"io"
 	"net/http"
-
-	httpgk "github.com/go-kit/kit/transport/http"
 )
 
 const (
-	// ContentTypeHeader is the 'Content-Type' header key.
+	// ContentTypeHeader is nhe 'Content-Type' header key.
 	ContentTypeHeader = "Content-Type"
 
 	// JSONContentType is the Content-Type header for JSON responses.
 	JSONContentType = "application/json; charset=utf-8"
 
 	// HTMLContentType is the Content-Type header for HTML responses.
 	HTMLContentType = "text/html; charset=utf-8"
-)
 
-// HTTPError is a specialized error that is returned be the Err method specified
-// in the ErrorResponse interface.
-type HTTPError struct {
-	// Code is the status code to set on the HTTP response.  Defaults to 500 if
-	// it is not set.
-	Code int
+	// PlainTextContentType is the Content-Type header for plain text responses.
+	PlainTextContentType = "text/plain; charset=utf-8"
+)
 
-	// Message is an optional message that can be returned to the
-	// client.  Defaults to the native http.StatusText message for the
-	// StatusCode() of the HTTPError.
-	Message string
+// ContentWriter is responsible writing the response body and communicating the
+// "Content-Type" of the response body.
+type ContentWriter interface {
+	// ContentType returns the value to set on the "Content-Type" header of the
+	// response.
+	ContentType() string
 
-	// Cause is an optional error that can be used to retain the error that
-	// led to us returning an HTTP error to the client.
-	Cause error
+	// WriteResponse takes the given response body and writes it to the given
+	// writer.
+	WriteResponse(io.Writer, interface{}) error
 }
 
-// Error returns the standard error string, this is not returned to the client.
-func (e HTTPError) Error() string {
-	return fmt.Sprintf(
-		"httpbp: http error with code %d, message %q and cause %v",
-		e.Code,
-		e.Message,
-		e.Cause,
-	)
+// ContentWriterFactory is the interface used by the handler to create new
+// ContentWriters when serving requests.
+type ContentWriterFactory func() ContentWriter
+
+// Response is an HTTP response that can be returned by a baseplate HTTP handler.
+//
+// Response is responsible for setting the values that are independant of the
+// body as well as the ContentWriter to write the response body.
+type Response interface {
```

Comment:
There's no way to set `ContentWriter` in `HTTPHandler`, does that mean for a microservice all handlers must share the same `ContentWriter` coming from the factory? e.g. you can't mix json and html?